### PR TITLE
Integrating vSwarm with InVitro

### PR DESCRIPTION
**Pull request by KarthikL1729**
## Summary

This PR is towards integrating the InVitro deployer and loader to the vSwarm benchmark suite. This is required as InVitro currently only uses dummy functions for deployments, and the vSwarm benchmark suite models realistic serverless loads which would help improve the versatility of InVitro.

## Implementation Notes :hammer_and_pick:

* Briefly outline the overall technical solution. If necessary, identify talking points where the reviewer's attention should be drawn to.

The deployer (Knative), invoker and driver have mainly been changed, and a new tool called the "mapper" has been added. The mapper reads the input trace files and maps the appropriate proxy functions from vSwarm based on the memory usage and function duration. Zip files for the vSwarm function yamls and their profiles have also been zipped and added as they are required to map and deploy the functions. vSwarm has been added as an option in the config files to allow the user to choose between using standard dummy functions from InVitro or the realistic functions from vSwarm.

The invoker has also been changed to include a new greeter client as the vSwarm functions require a SayHello gRPC call to run as opposed to the executeRPC function call. The recording of the successful and failed requests has also been modified to only count the requests after the warmup duration is done.

In summary, this PR adds a new tool (`tools/mapper`), zipped yamls for the vSwarm functions (`workloads/container/vSwarm_yamls.tar.gz`), modifies the `driver` and `cmd/loader.go` files to integrate vSwarm as an option.

An automatic deployer script has also been added to allow for multiple run sweeps.

## External Dependencies :four_leaf_clover:

* N/A

## Breaking API Changes :warning:

* N/A

*Simply specify none (N/A) if not applicable.*

### CODE REVIEW

**Review by leokondrashov — COMMENTED** [review]
Adding 12k lines????
Please consider removing them from the PR or explaining why you cannot go without them. It's impossible to review such a big PR: github webUI freezes from time to time.

Please update the PR description to be more verbose on what, how, and why you do the things in the PR, as it is specified in the PR template.

File: .gitattributes
Comment (leokondrashov) [inline .gitattributes:1]: Move gitattributes to respective folders

File: cmd/loader.go [hunk 1]
```diff
@@ -144,7 +144,11 @@ func parseIATDistribution(cfg *config.LoaderConfiguration) (common.IatDistributi
 func parseYAMLSpecification(cfg *config.LoaderConfiguration) string {
 	switch cfg.YAMLSelector {
 	case "container":
-		return "workloads/container/trace_func_go.yaml"
+		if cfg.VSwarm {
```
Comment (leokondrashov) [inline cmd/loader.go:147]: How would loader react if we try to run vSwarm with Firecracker?

File: cmd/loader.go [hunk 2]
```diff
@@ -174,7 +178,7 @@ func runTraceMode(cfg *config.LoaderConfiguration, readIATFromFile bool, justGen
 	yamlPath := parseYAMLSpecification(cfg)
 
 	traceParser := trace.NewAzureParser(cfg.TracePath, durationToParse)
-	functions := traceParser.Parse(cfg.Platform)
+	functions := traceParser.Parse(cfg.Platform)		// Indexed by hashfunction names
```
Comment (leokondrashov) [inline cmd/loader.go:181]: I don't see any indexing here. What does this comment means then?

File: pkg/common/constants.go [hunk 3]
```diff
@@ -83,7 +83,7 @@ const (
 	FailedWarnThreshold = 0.3
 	// FailedTerminateThreshold Terminate experiment if the percentage of failed invocations (e.g., connection timeouts,
 	// function timeouts) is greater than this threshold
-	FailedTerminateThreshold = 0.5
+	FailedTerminateThreshold = 1.1
```
Comment (leokondrashov) [inline pkg/common/constants.go:86]: Why do you change it here?

File: pkg/driver/clients/grpc_client.go [hunk 4]
```diff
@@ -29,10 +29,12 @@ import (
 	"go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc"
 	"strings"
 	"time"
-
+	"github.com/google/uuid"
```
Comment (leokondrashov) [inline pkg/driver/clients/grpc_client.go:32]: Please run through gofmt before the push

File: pkg/driver/trace_driver.go [hunk 5]
```diff
@@ -74,7 +74,7 @@ func NewDriver(driverConfig *config.Configuration) *Driver {
 // HELPER METHODS
 // ///////////////////////////////////////
 func (d *Driver) outputFilename(name string) string {
-	return fmt.Sprintf("%s_%s_%d.csv", d.Configuration.LoaderConfiguration.OutputPathPrefix, name, d.Configuration.TraceDuration)
+	return fmt.Sprintf("%s_%s_%d_%d.csv", d.Configuration.LoaderConfiguration.OutputPathPrefix, name, d.Configuration.TraceDuration, len(d.Configuration.Functions))
```
Comment (leokondrashov) [inline pkg/driver/trace_driver.go:77]: Why do you need to change the filename format?

**Review by leokondrashov — CHANGES_REQUESTED** [review]
This is a result of my brief look at the code.

Please also update the tests that fail.

**Comment by leokondrashov** [comment]
Please sign off the commits (read contributor's guide in vhive)

File: cmd/loader.go (hunk 2 above)
Comment (KarthikL1729) [reply cmd/loader.go:181]: This was more for my understanding when modifying the code, that the hashfunction is what I would use to address the function that the specific hash function is mapped to. Will remove it.

File: cmd/loader.go (hunk 1 above)
Comment (KarthikL1729) [reply cmd/loader.go:147]: We have not tested with firecracker since I have not modified anything that relates to firecracker specifically. For now, it is only for Knative and so I think adding a condition for vSwarm in the firecracker block to exit and throw an error would be okay?

File: cmd/loader.go (hunk 2 above)
Comment (leokondrashov) [reply cmd/loader.go:181]: You don't use it for indexing. `functions` is a flat array.

File: pkg/common/constants.go (hunk 3 above)
Comment (KarthikL1729) [reply pkg/common/constants.go:86]: I changed this for testing, but I was not sure why we want the drivers to terminate if the failure rate is over 50%. Shouldn't it be okay to run them even if the rate of failure is above 50%?

File: pkg/driver/trace_driver.go (hunk 5 above)
Comment (KarthikL1729) [reply pkg/driver/trace_driver.go:77]: I changed it because of the automated deployer, which runs a sweep over the number of functions with the same durations. So without this, the same file would be overwritten multiple times.

File: tools/mapper/trace_parser.py
[the remark this answers is not in the document]
Comment (KarthikL1729) [reply tools/mapper/trace_parser.py:1]: The reason this file exists is to convert the csv into Ordered dicts and for having a uniform naming scheme for the columns. We need Ordered dicts for the mapping process, since the memory and duration traces need to be in sync; and the uniform naming scheme is as the trace itself has varying names for the percentiles between durations and memory.

File: cmd/loader.go (hunk 1 above)
Comment (leokondrashov) [reply cmd/loader.go:148]: Yes, that is sensible. Please also update docs which modes are supported, which are not.

File: tools/mapper/trace_parser.py
[the remark this answers is not in the document]
Comment (leokondrashov) [reply tools/mapper/trace_parser.py:1]: Do you know that you can rename the columns in the DataFrame? I don't see your explanation as a valid reason to add another parser to the code.

**Comment by leokondrashov** [comment]
When updating the branch, please use rebase, not the merge option.

File: pkg/driver/trace_driver.go (hunk 5 above)
Comment (leokondrashov) [reply pkg/driver/trace_driver.go:78]: Overall, is the automatic deployer an integral part of this PR? Can the automatic deployer be changed so as not to overwrite the file? I ask this because that messes up the interface with the invitro (which you didn't declare in the PR description BTW) that my analysis code relies on. So, change it only if you actually need it to be different.

File: pkg/driver/clients/grpc_client.go [hunk 6]
```diff
@@ -87,38 +89,63 @@ func (i *grpcInvoker) Invoke(function *common.Function, runtimeSpec *common.Runt
...
+			return false, record
+		}
+		record.ResponseTime = time.Since(start).Microseconds()
+		record.ActualDuration = uint32(record.ResponseTime)
+		record.Instance = extractInstanceName(response.GetMessage())
+		if strings.HasPrefix(response.GetMessage(), "FAILURE - mem_alloc") {
+			record.MemoryAllocationTimeout = true
+		} else {
+			record.ActualMemoryUsage = common.Kib2Mib(0) //common.Kib2Mib(response.MemoryUsageInKb)
+		}
```
Comment (leokondrashov) [inline pkg/driver/clients/grpc_client.go:146]: Extract this code into a new vSwarmGrpcInvoker instead of making a special case for the grpcInvoker

File: pkg/driver/deployment/knative.go [hunk 7]
```diff
@@ -83,11 +87,63 @@ func knativeDeploySingleFunction(function *common.Function, yamlPath string, isP
 		// second, then round to an integer as that is what the knative config expects
 	}
 
+	// Default is the function name, but if the function is a vSwarm function, the proxy name is used
+	proxyName := function.Name
+
+	if vSwarm {
+		// Read and unmarshal the mapper output file into a map
+
+		mapperFile, err := os.ReadFile("tools/mapper/output.json")
```
Comment (leokondrashov) [inline pkg/driver/deployment/knative.go:96]: Can you make it without hard-coding the path to the file? For example, it might be more useful to move mapper's output to the trace directory instead of using the global file for the mappings.

File: cmd/loader.go [hunk 8]
```diff
@@ -145,8 +145,15 @@ func parseIATDistribution(cfg *config.LoaderConfiguration) (common.IatDistributi
 func parseYAMLSpecification(cfg *config.LoaderConfiguration) string {
 	switch cfg.YAMLSelector {
 	case "container":
-		return "workloads/container/trace_func_go.yaml"
+		if cfg.VSwarm {
+			return "workloads/container/yamls/"
```
Comment (leokondrashov) [inline cmd/loader.go:149]: Rename this directory to vSwarm yamls

File: pkg/driver/deployment/knative.go [hunk 9]
```diff
@@ -83,11 +87,63 @@ func knativeDeploySingleFunction(function *common.Function, yamlPath string, isP
...
+		proxyName = mapperOutput[function.HashFunction]["proxy-function"]
+		log.Info("Proxy name: ", proxyName)
+		// Read the deployment info file for yaml locations and predeployment commands if any
+
+		deploymentInfoFile, err := os.ReadFile(yamlPath + "deploy_info.json")
+		if err != nil {
+			log.Warn("No deployment info file")
+		}
+
+		var deploymentInfo map[string]map[string]interface{}
```
Comment (leokondrashov) [inline pkg/driver/deployment/knative.go:116]: Why do you use this data structure? Can you use sensible structs for that?

File: pkg/driver/deployment/knative.go [hunk 10]
```diff
@@ -83,11 +87,63 @@ func knativeDeploySingleFunction(function *common.Function, yamlPath string, isP
...
+		if err != nil {
+			log.Warn("Failed to unmarshal mapper output file")
+		}
+
+		// Find the proxy name for the function
+		proxyName = mapperOutput[function.HashFunction]["proxy-function"]
+		log.Info("Proxy name: ", proxyName)
+		// Read the deployment info file for yaml locations and predeployment commands if any
+
+		deploymentInfoFile, err := os.ReadFile(yamlPath + "deploy_info.json")
```
Comment (leokondrashov) [inline pkg/driver/deployment/knative.go:111]: This file seems important for understanding the code. Can you pull it out of the tar.gz?

File: tools/mapper/trace_parser.py
[the remark this answers is not in the document]
Comment (KarthikL1729) [reply tools/mapper/trace_parser.py:1]: The name change is not the main reason. The main reason is populating an ordered dict, to make sure that the durations and memory traces are preserved in order.

File: pkg/driver/deployment/knative.go (hunk 9 above)
Comment (KarthikL1729) [reply pkg/driver/deployment/knative.go:116]: This data structure is to unmarshal the json file that has the predeployment commands that are required for deploying functions that require databases.

Comment (leokondrashov) [reply pkg/driver/deployment/knative.go:116]: Can you use structure of structures instead of map of maps?

File: pkg/driver/trace_driver.go (hunk 5 above)
Comment (KarthikL1729) [reply pkg/driver/trace_driver.go:78]: Will change it back and try to figure something else out for the automatic deployer.

File: pkg/driver/deployment/knative.go (hunk 9 above)
Comment (KarthikL1729) [reply pkg/driver/deployment/knative.go:116]: Should I define a new struct for this? I thought this is a pretty simple interface to unmarshal into and that adding a specific struct for this would complicate things.

**Comment by leokondrashov** [comment]
Will be splitted into parts and reviewed separately
